internal/services/v1/experimental.go

```diff
@@ -386,6 +406,153 @@ func (es *experimentalServer) BulkExportRelationships(
 	return nil
 }
 
+func (es *experimentalServer) BulkCheckPermission(ctx context.Context, req *v1.BulkCheckPermissionRequest) (*v1.BulkCheckPermissionResponse, error) {
+	atRevision, checkedAt, err := consistency.RevisionFromContext(ctx)
+	if err != nil {
+		return nil, es.rewriteError(ctx, err)
+	}
+
+	// Identify checks with same permission+subject over different resources and group them. This is doable because
+	// the dispatching system already internally supports this kind of batching for performance.
+	groupingParams := groupingParameters{
+		atRevision:           atRevision,
+		maxCaveatContextSize: es.maxCaveatContextSize,
+		maximumAPIDepth:      es.maximumAPIDepth,
+	}
+	groupedItems, err := groupItems(ctx, groupingParams, req.Items)
+	if err != nil {
+		return nil, es.rewriteError(ctx, err)
+	}
+
+	bulkResponseMutex := sync.Mutex{}
+	resp := &v1.BulkCheckPermissionResponse{CheckedAt: checkedAt}
+	tr := taskrunner.NewPreloadedTaskRunner(ctx, es.bulkCheckMaxConcurrency, len(groupedItems))
+
+	appendResultsForError := func(group groupedCheckParameters, err error) error {
+		bulkResponseMutex.Lock()
+		defer bulkResponseMutex.Unlock()
+
+		rewritten := es.rewriteError(ctx, err)
+		statusResp, ok := status.FromError(rewritten)
+		if !ok {
+			// If error is not a gRPC Status, fail the entire bulk check request.
+			return err
+		}
+
+		for _, resourceID := range group.resourceIDs {
+			reqItem, err := requestItemFromResourceAndParameters(group.params, resourceID)
+			if err != nil {
+				return es.rewriteError(ctx, err)
+			}
+
+			resp.Pairs = append(resp.Pairs, &v1.BulkCheckPermissionPair{
+				Request: reqItem,
+				Response: &v1.BulkCheckPermissionPair_Error{
+					Error: statusResp.Proto(),
+				},
+			})
+		}
+
+		return nil
+	}
+
+	appendResultsForCheck := func(params computed.CheckParameters, resourceIDs []string, metadata *dispatchv1.ResponseMeta, results map[string]*dispatchv1.ResourceCheckResult) error {
```

Comment:
metadata isn't actually being used here

Reply:
Good catch; fixed

Reply:
I don't see it fixed yet

Reply:
https://github.com/authzed/spicedb/pull/1453/files#diff-46f0df22900bde355de03fe5c3e40b0dd39ad7332bd23b41a5dc8079e526d0afR482